src/algebra/module/polynomial.lean

```diff
@@ -0,0 +1,144 @@
+/-
+Copyright (c) 2022 Pierre-Alexandre Bazin. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Pierre-Alexandre Bazin
+-/
+import data.polynomial
+
+/-!
+# Modules over `R[X]`
+
+## Main definitions
+
+* `module_with_End E` : A type synonym for a `R`-module which explicitly mentions a given
+  endomorphism `E`. It naturally comes with a `R[X]`-module structure, where the action of `X` is
+  given by `E`.
+
+## Main statements
+
+* `module_with_End.polynomial_linear_equiv_semiconj` : equivalence between `R[X]`-linear maps and
+  `R`-linear maps compatible with the action of `X`. See also
+  `module_with_End.polynomial_lequiv_equiv_semiconj` for the restriction to linear isomorphisms.
+
+## Tags
+
+Module, polynomial
+-/
+
+
+variables {R M : Type*} [comm_semiring R] [add_comm_monoid M] [module R M]
+open_locale polynomial
+
+
+open polynomial module
+
+/--A type synonym for a `R`-module which explicitly mentions a given endomorphism.-/
+@[nolint unused_arguments has_inhabited_instance]
+def module_with_End (E : End R M) := M
+
+namespace module_with_End
+variables {E : End R M}
+
+instance : add_comm_monoid (module_with_End E) := (infer_instance : add_comm_monoid M)
+instance : module R (module_with_End E) := (infer_instance : module R M)
+/--The type conversion `M → module_with_End E`, as a `R`-linear isomorphism.-/
+def of_module (E : End R M) : M ≃ₗ[R] module_with_End E := linear_equiv.refl R M
+@[simp] lemma of_module_apply (u : M) : of_module E u = u := rfl
```

Comment:
This lemma defeats the whole point of having `of_module`.
What goes wrong if you remove it?

Reply:
If you end up in a position like that, then you've already taken a wrong turn. What type is `f` in that example? Is it a function from `M` or from `module_with_End E`?

Reply:
I was sure I commented with corrected versions of `X_smul`, but github seems to have discarded them...